extensions/amp-consent/0.1/amp-consent.js

```diff
@@ -262,25 +267,46 @@ export class AmpConsent extends AMP.BaseElement {
    * Register a list of user action functions
    */
   enableInteractions_() {
-    this.registerAction('accept', () => {
-      this.handleAction_(ACTION_TYPE.ACCEPT);
+    this.registerAction('accept', (invocation) => {
+      this.handleClosingUIAction_(ACTION_TYPE.ACCEPT, invocation);
     });
 
-    this.registerAction('reject', () => {
-      this.handleAction_(ACTION_TYPE.REJECT);
+    this.registerAction('reject', (invocation) => {
+      this.handleClosingUIAction_(ACTION_TYPE.REJECT, invocation);
     });
 
     this.registerAction('dismiss', () => {
-      this.handleAction_(ACTION_TYPE.DISMISS);
+      this.handleClosingUIAction_(ACTION_TYPE.DISMISS);
     });
 
+    if (this.isGranularConsentExperimentOn_) {
+      this.registerAction('setPurpose', (invocation) => {
+        this.handleSetPurpose_(invocation);
+      });
+    }
+
     this.registerAction('prompt', (invocation) =>
       this.handleReprompt_(invocation)
     );
 
     this.enableExternalInteractions_();
   }
 
+  /**
+   * For actions that close the PromptUI, validate state
+   * and do some preprocessing, then handle action.
+   * @param {string} action
+   * @param {../../../src/service/action-impl.ActionInvocation=} opt_invocation
+   */
+  handleClosingUIAction_(action, opt_invocation) {
+    if (this.isReadyToHandleAction_()) {
```

Comment:
bail early e.g.
```
if (!this.isReadyToHandleAction_()) {
  return;
}
// do other things
```
This also helps with readability since the entire function body is not within the if block.